# This script is to update table "consensus_sequence" sequencing qa columns with values in V-pipe-generated qa.csv
# Values in database will be current to values in qa.csv.

#' @param db_connection
#' @param qa_data_path Path to qa.csv generated by V-Pipe.
#' @param append_new_rows Should always be F, because check for unique sample_names removed, update done now based on key (sample_name, sequencing_batch).
import_sequencing_qa <- function (db_connection, qa_data_path, append_new_rows = F) {
  TBL_NAME <- "consensus_sequence"
  cols_of_interest <- c(
    "sample_name", "coverage", "r1_basequal", "r2_basequal", "rejreads", 
    "alnreads", "insertsize", "consensus_n", "consensus_lcbases", 
    "sequencing_center")
  key_col <- c("sample_name", "sequencing_batch")

  # Load data
  data <- read.delim(file = qa_data_path, sep = ",", stringsAsFactors = F)

  # Format data
  sequencing_qa <- data %>%
    rename("sample_name" = "sample",
           "sequencing_center" = "lab",
           "sequencing_batch" = "batch",
           "coverage" = "coverage_mean",
           "r1_basequal" = "fastqc_r1_basequal",
           "r2_basequal" = "fastqc_r2_basequal",
           "consensus_n" = "consensus_N",
           "alnreads_n" = "alnreads",
           "insertsize" = "bwa_insert_mean",
           "consensus_lcbases" = "consensus_lower") %>%
    mutate(rejreads = 100 - goodpairs * 2 * 100 / (input_r1 + input_r2),  # percentage of raw reads rejected based on Q-scores
           alnreads = alnreads_n * 50 / goodpairs)  # percentage of kept reads aligned to reference

  table_spec <- parse_table_specification(
    table_name = TBL_NAME, db_connection = db_connection)
  cols_to_update <- cols_of_interest[cols_of_interest != key_col]

  # Update table with data
  update_table(
    table_name = TBL_NAME, 
    new_table = sequencing_qa,
    con = db_connection, 
    append_new_rows = F,
    cols_to_update = cols_to_update,
    key_col = key_col,
    table_spec = table_spec)
}
